### PR TITLE
fix(js) Implement Numeric Separator proposal

### PR DESCRIPTION
**Pull request by aduh95**
This PR implements the Numeric Separator proposal, currently stage 3 of TC39 process, and supported by all major browsers.

Limitations:
- the proposal doesn't allow multiple adjacent separator. Because it makes the regex overcomplicated, I've ignored it.
- Number literals with leading zero (which represents either decimal or octal literal) are not supported by this PR. Worth noting they are illegal in strict mode (incl. ES modules), so I believe it makes sense to not highlight them.

Refs: https://github.com/tc39/proposal-numeric-separator
Refs: https://tc39.es/ecma262/#sec-literals-numeric-literals

### CODE REVIEW

**Comment by yyyc514** [comment]
I only took a glance.  This may have to wait until after the JS/TS merge I've been working on.  But even though I don't think it'd be a big deal to reintegrate it after.

I'll try to take a deeper look next week.

File: src/languages/javascript.js
Comment (joshgoebel) [inline src/languages/javascript.js:46]: Is there a strong reason you have included `-`?  Typically we do not do this because it will break math expressions like `3-2`... there is no simple way to detect that it's truly a unary expression or not...

Sorry for the long delay on this... finally polished up my TS/JS work and hope to have it merged soon so I was taking another look at this.

Comment (aduh95) [reply src/languages/javascript.js:46]: > Is there a strong reason you have included `-`?

TBH I have no recollection what I was trying to achieve here. Once the other PR gets merge, I'll try to come back at this one to see if that's OK to just remove that dash.

Comment (joshgoebel) [reply src/languages/javascript.js:46]: > Once the other PR gets merge,

Merged. :)

**Comment by aduh95** [comment]
I've rebased and included a second commit to remove the `-` from the regex. I guess I should include a few more test cases to make explicit what is the policy of `highlight.js` when it comes to dash and numbers:

```js
let num = -3; // `-` _should_ be included, but is not anymore
let num = - 3; // Who writes JS like that??
let diff = 5 -3; // Who writes JS like that??
let diff = 5- 3; // Who writes JS like that??
let diff = 5 - 3; // yep, pretty classic
let diff = 5-3; // minified version

**Comment by joshgoebel** [comment]
> EDIT: Let's try if my original commit is able to handle these cases before dismissing it.

Nope.

File: test/markup/javascript/keywords.expect.txt
Comment (aduh95) [inline test/markup/javascript/keywords.expect.txt:3]: That looks like a regression, but maybe it's fine.

Comment (joshgoebel) [reply test/markup/javascript/keywords.expect.txt:3]:
```
<span class="hljs-keyword">let</span> diff = <span class="hljs-number">5</span><span class="hljs-number">-3</span>;
```

Screen Shot 2020-10-03 at 10 44 05 AM

I'd say actually it's a FIX - just we didn't have good numeric tests in the past. :-)  C_NUMBER includes `-?` in the regex, which as I said is problematic... it's probably something we should consider changing (for exactly this reason), but that's a separate issue.

When look behind is added we could add `-` back when it immediately follows certain things like `(` or `=`, but that's also another day.

File: src/languages/javascript.js
Comment (joshgoebel) [inline src/languages/javascript.js:78]: Is there a reason we have no `\b` here?

File: src/languages/javascript.js
Comment (joshgoebel) [inline src/languages/javascript.js:85]: Is sequencing off here?  Isn't it `[.]?n?`?  `0n.` would not be correct, would it?

Comment (joshgoebel) [reply src/languages/javascript.js:85]: Oh, it's either... :-) So then I think this is ok since `.0n` does not seem valid.

**Comment by joshgoebel** [comment]
Looking good.  Left one comment.  Also please add a CHANGES.md entry. :-)

Almost there. :)

File: src/languages/javascript.js
Comment (aduh95) [reply src/languages/javascript.js:78]: Probably not, let's try with it

```suggestion
      { begin: regex.concat(/\b0?\./, decimalDigits, regex.optional(exponentPart)) }, // Decimal literals between 0 and 1
```

Comment (aduh95) [reply src/languages/javascript.js]: Hum so now the tests don't pass…

**Review by joshgoebel — APPROVED** [review]

File: src/languages/javascript.js
Comment (joshgoebel) [reply src/languages/javascript.js]: Does the \b need to be grouped with just the 0?  I don't think there is a boundary between say a " " and a "."?

File: src/languages/javascript.js
Comment (joshgoebel) [inline src/languages/javascript.js:78]:
```suggestion
      { begin: regex.concat(/(\b0)?\./, decimalDigits, regex.optional(exponentPart)) }, // Decimal literals between 0 and 1
```

**Review by joshgoebel — APPROVED** [review]

**Comment by joshgoebel** [comment]
@aduh95 Thanks for all the work on this, and sorry for the lag.

**Pull request merged**
